spatial/curve/hilbert.go

```diff
@@ -0,0 +1,236 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package curve
+
+// Hilbert2D is a 2-dimensional Hilbert curve.
+type Hilbert2D struct{ Order int }
+
+// Dims returns the spatial dimensions of the curve, which is {2ᵏ, 2ᵏ}, where k
+// is the order.
+func (h Hilbert2D) Dims() []int { return []int{1 << h.Order, 1 << h.Order} }
+
+// Len returns the length of the curve, which is 2ⁿᵏ, where n is the dimension
+// (2) and k is the order.
+//
+// Len will overflow on a 32-bit architecture if the order is ≥ 16.
+func (h Hilbert2D) Len() int { return 1 << (2 * h.Order) }
+
+func (h Hilbert2D) rot(n int, v []int, d int) {
+	switch d {
+	case 0:
+		swap{0, 1}.do(n, v)
+	case 3:
+		flip{0, 1}.do(n, v)
+	}
+}
+
+// Curve returns the curve coordinate of v. Curve modifies v.
+//
+// Curve will overflow on a 32-bit architecture if the order is ≥ 16.
+func (h Hilbert2D) Curve(v []int) int {
+	var d int
```

Comment:
> Should there be a check that `len(v) == 2`?

My thought was more or less, if `v` is too short, it will panic anyways, and if `v` is longer than necessary *shrug* hopefully the user knows what they're doing.

> Are there any restrictions on the components of v?

The values produced by `Space` should be `0 ≤ v < 2ᵏ`. If a value outside of that range is passed to `Curve`, I'm guessing it would give the same result as calling it with each value modulo 2ᵏ. Curve takes N bits at a time, so I think it would ignore everything except the last k bits of each value.